### DISCUSSION TITLE
OSX Display tearing

### DISCUSSION THREAD

**Discussion by IcemarkUK** [axmolengine/axmol#1684] [Q&A] [answered]
It's a difficult one to show, as video doesn't capture it.

Compared to Cocos2d-x v4 all large movement (transitions etc) seems to show a lot of tearing. Most noticeably in game when scrolling a full window of a TMX layer. Happens in Windowed or Fullscreen mode. I don't see any tearing on Cocos2d-x

This could be connected to [my issue with full screen](https://github.com/axmolengine/axmol/discussions/1667) as my iMac is 5120x2880 but running at 2880x1620

I wondered if there is a problem with too big frame buffer, therefore doing too much work, or refresh/fps/vsync issue

**Comment by rh101** [comment]
Are you able to use any profiling tools on the Mac? Finding any code hotspots, excessive graphic related calls etc. may help.

**Comment by rh101** [comment]
I didn't notice any screen tearing while running it at 2560x1440.  Is there a specific part of the app I should be looking at? I was in the map screen, moving it around.

**Reply by IcemarkUK** [reply]
It's most obvious when scrolling around the full map, but I see it on all the transitions and when moving the main view around (dragging the display window). But I do suspect it's something to do with apple retina on my iMac and that my display is 5k

If you check map_panel_detailed::Init

```
    if ( mr->config->debug_map ) {
        mapBuilder->setFlags(mapflags::debug_map);
        mapBuilder->setFlags(mapflags::show_all_characters);
        mapBuilder->setFlags(mapflags::show_all_critters);
    }
```

With debug map on you can scroll the whole map.

**Reply by IcemarkUK** [reply, marked as answer]
It appears to be this...

https://github.com/axmolengine/axmol/blob/bee2d6222d6d7095b9be4c07c58b4e842b19e450/core/platform/GLViewImpl.cpp#L524-L533

If I set layer.displaySyncEnabled  = true then the problem goes away - which makes sense.

**Reply by rh101** [reply]
I'm glad you found the issue.  This shouldn't require any game engine changes, since it's just a matter of setting the `vsync` value via `GLView::setGLContextAttrs` (in AppDelegate), or if done dynamically, then recreating the view if required.

**Reply by IcemarkUK** [reply]
Seems like there should be a more formal way of doing this. I've not seen anything in engine or documentation about vsync. And it also looks like a change of behaviour. I don't think its set in Cocos2d-x - so I think it must default to true as opposed to false.

**Reply by rh101** [reply]
Cocos2d-x didn't have that section of code to set the `layer.displaySyncEnabled`, neither did it have the `vsync` member in `GLContextAttrs`, as you can see here:
```
// Cocos2d-x v4 GLContextAttrs
struct GLContextAttrs
{
    int redBits;
    int greenBits;
    int blueBits;
    int alphaBits;
    int depthBits;
    int stencilBits;
    int multisamplingCount;
};
```

The reason it was working in Cocos2d-x was purely a fluke, given that the default value of `displaySyncEnabled` is `true`:
https://developer.apple.com/documentation/quartzcore/cametallayer/2887087-displaysyncenabled?language=objc

I wasn't aware of the `vsync` addition in Axmol, and never had a reason to look into it, but at least we know about it now, and can make sure it's documented.  It's actually a good addition, since the game can switch it on or off at runtime, even having it as an option the user can change.

It would be as straightforward as:
```
auto glContextAttrs = GLView::getGLContextAttrs();
glContextAttrs.vsync = true|false;
GLView::setGLContextAttrs(glContextAttrs);
```
Then call `GLViewImpl::createWithFullScreen` or `GLViewImpl::createWithRect` to recreate the view.
